server.py

```diff
@@ -0,0 +1,82 @@
+#! /usr/bin/env python3
+
+# Author: Thomas Gruber
+# Version: 0.8, 11/06/18
+
+from http.server import HTTPServer, BaseHTTPRequestHandler
+
+import io
+import json
+import time
+
+
+class ControlHandler (BaseHTTPRequestHandler):
+    SERVER_PATH = '/terradrone'
+    CHAR_ENCODING = 'utf-8'
+
+    # commands should correspond to those written in Android application
+    CMD_ERROR                                                   = 0
+    CMD_TEST                                                    = 1
+    CMD_FORWARD_START                                           = 100
+    CMD_FORWARD_STOP                                            = 101
+    CMD_BACKWARD_START                                          = 102
+    CMD_BACKWARD_STOP                                           = 103
+    CMD_ROTATE_RIGHT_START                                      = 104
+    CMD_ROTATE_RIGHT_STOP                                       = 105
+    CMD_ROTATE_LEFT_START                                       = 106
+    CMD_ROTATE_LEFT_STOP                                        = 107
+    CMD_COLLECT                                                 = 200
+
+
+    def do_GET(self):
+        # deny request
+        self.send_error(418)
+
+
+    def do_POST(self):
+        content_length = int(self.headers['Content-Length'])
+        json_bytes = self.rfile.read(content_length).decode(self.CHAR_ENCODING)
+
+        request = json.loads(json_bytes)
+        request_type = request['command']
+
+        # only allow requests with correct path (weak security feature)
+        if self.path == self.SERVER_PATH:
+            if request_type == self.CMD_TEST:
+                response = {
+                    'time': int(round(time.time() * 1e3)),
+                    'command': self.CMD_TEST
+                }
+
+                json_response = json.dumps(response).encode(self.CHAR_ENCODING)
+
+                content_buffer = io.BytesIO()
+                response_length = content_buffer.write(json_response)
+
+                self.send_response(202)
+                self.send_header('Content-Length', response_length)
+                self.send_header('Content-Type', 'application/json; charset=' + self.CHAR_ENCODING)
+                self.send_header('Connection', 'close')
+                self.end_headers()
+
+                self.wfile.write(content_buffer.getvalue())
+                content_buffer.close()
+            else:
+                # invalid command type
+                self.send_error(400)
+        else:
+            print(self.path)
+            self.send_error(401)
+
+        # default to closing all connections
+        self.close_connection = True
+
+
+def main():
+    # must be ran with 'sudo' to run on port 80
+    httpd = HTTPServer(('', 80), ControlHandler)
```

Comment:
This is for Apache, correct? Did we have to install Apache on the Pi for this to work?

Reply:
I did not have to install any extra libraries. Either it came preinstalled on the Pi or with the Python installation.